File: 프로그래머스/2/12952. N－Queen/N－Queen.py
def solution(n):
    columnCheck = [False] * n
    daegakCheck1 = [False] * (2 * n + 1)
    daegakCheck2 = [False] * (2 * n + 1)

    answer = 0

    def check(y, x):
        if columnCheck[x] or daegakCheck1[y + x] or daegakCheck2[n + x - y]:
            return False

        return True

    def mark(y, x):
        columnCheck[x] = True
        daegakCheck1[y + x] = True
        daegakCheck2[n + x - y] = True

    def unmark(y, x):
        columnCheck[x] = False
        daegakCheck1[y + x] = False
        daegakCheck2[n + x - y] = False

    def solve(y):
        nonlocal answer
        if y == n:
            answer += 1
            return

        for x in range(n):
            if check(y, x):
                mark(y, x)
                solve(y + 1)
                unmark(y, x)

    solve(0)

    return answer

File: 백준/Gold/9663. N－Queen/N－Queen.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {

    static int n, answer;
    static boolean[] columnCheck, daegakCheck1, daegakCheck2;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        n = Integer.parseInt(br.readLine());
        columnCheck = new boolean[n];
        daegakCheck1 = new boolean[2 * n + 1];
        daegakCheck2 = new boolean[2 * n + 1];

        answer = 0;

        solve(0);

        System.out.println(answer);

    }

    private static void solve(int y) {
        if (y == n) {
            answer++;
            return;
        }

        for (int x = 0; x < n; x++) {
            if (check(y, x)) {
                mark(y, x, true);
                solve(y + 1);
                mark(y, x, false);
            }
        }
    }

    private static boolean check(int y, int x) {
        if (columnCheck[x] || daegakCheck1[x + y] || daegakCheck2[n + x - y]) {
            return false;
        }
        return true;
    }

    private static void mark(int y, int x, boolean tf) {
        columnCheck[x] = tf;
        daegakCheck1[x + y] = tf;
        daegakCheck2[n + x - y] = tf;
    }
}